contractsV5/KyberHintHandler.sol

```diff
@@ -216,84 +247,103 @@ contract KyberHintParser is IKyberHint, Utils {
+        } else if (opcodeHash == keccak256(encodeOpcode(TradeType.MaskOut))) {
+            reservesHint.tradeType = TradeType.MaskOut;
+            (indexToContinueFrom) = decodeReservesFromHint(false, hint, reservesHint, isExternal, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth, isExternal);
+        } else if (opcodeHash == keccak256(encodeOpcode(TradeType.Split))) {
+            reservesHint.tradeType = TradeType.Split;
+            (indexToContinueFrom) = decodeReservesFromHint(true, hint, reservesHint, isExternal, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth, isExternal);
+        } else {            
+            reservesHint.invalidHint = true;
         }
     }
 
     function decodeReservesFromHint(
         bool isTokenSplit,
         bytes memory hint,
+        ReservesHint memory reservesHint,
+        bool isExternal,
         uint indexToContinueFrom
     )
         internal
         view
         returns (
-            IKyberReserve[] memory,
-            uint[] memory,
             uint
         )
     {
        uint reservesLength = hint.toUint8(indexToContinueFrom);
        uint bpsSoFar;
        uint[] memory splitValuesBps;
-
        IKyberReserve[] memory addresses = new IKyberReserve[](reservesLength);
+       bytes8[] memory reserveIds = new bytes8[](reservesLength);
+
        if (isTokenSplit) {
            splitValuesBps = new uint[](reservesLength);
        } else {
            splitValuesBps = new uint[](1);
            splitValuesBps[0] = BPS;
        }
+
        indexToContinueFrom++;
 
-       for (uint i = 0; i < reservesLength; i++) {           
-           addresses[i] = IKyberReserve(convertReserveIdToAddress(hint.slice(indexToContinueFrom, RESERVE_ID_LENGTH).toBytes8(0)));
+       for (uint i = 0; i < reservesLength; i++) {
+           reserveIds[i] = hint.slice(indexToContinueFrom, RESERVE_ID_LENGTH).toBytes8(0);
+
+           if (!isExternal) {
+               addresses[i] = IKyberReserve(convertReserveIdToAddress(reserveIds[i]));
+           }
+
            indexToContinueFrom += RESERVE_ID_LENGTH;
+
            if (isTokenSplit) {
                splitValuesBps[i] = uint(hint.toUint16(indexToContinueFrom));
                bpsSoFar += splitValuesBps[i];
                indexToContinueFrom += 2;
            }
        }
+
        require((bpsSoFar == BPS) || (bpsSoFar == 0), "bps do not sum to 10000");
-       return (addresses, splitValuesBps, indexToContinueFrom);
+
+       reservesHint.splitValuesBps = splitValuesBps;
+       if (isExternal) {
+           reservesHint.reserveIds = reserveIds;
+       } else {
+           reservesHint.addresses = addresses;
+       }
+
+       return (indexToContinueFrom);
     }
 
-    function parseOpcode(TradeType h) internal pure returns (bytes memory) {
+    function encodeOpcode(TradeType h) internal pure returns (bytes memory) {
```

Comment:
what is h?
can it be type or something of that sort.

Reply:
Will change it to type.

Reply:
Changed.